iocore/net/SSLSessionCache.h

```diff
@@ -187,16 +187,16 @@ class SSLOriginSession
 {
 public:
   std::string key;
-  Ptr<IOBufferData> asn1_data; /* this is the ASN1 representation of the SSL_CTX */
-  size_t len_asn1_data;
-  Ptr<IOBufferData> extra_data;
+  SSL_SESSION *session;
+  ssl_curve_id curve_id;
 
-  SSLOriginSession(const std::string &lookup_key, const Ptr<IOBufferData> &ssl_asn1_data, size_t len_asn1,
-                   Ptr<IOBufferData> &exdata)
-    : key(lookup_key), asn1_data(ssl_asn1_data), len_asn1_data(len_asn1), extra_data(exdata)
+  SSLOriginSession(const std::string &lookup_key, SSL_SESSION *sess, ssl_curve_id curve)
+    : key(lookup_key), session(sess), curve_id(curve)
   {
   }
 
+  ~SSLOriginSession() { SSL_SESSION_free(session); }
```

Comment:
Not really, this is freeing the session when it's being replaced or removed from the cache, as we're not doing the `i2d_SSL_SESSION` conversion and returning the correct return value through the new session callback.

Reply:
The reason for the leak is because when we do the `i2d_SSL_SESSION` for the server sessions, we always return 0 so the reference count for the actual OpenSSL's session object is decreased after returning from the callback, and we do some manual cleanups in our NetVC if we convert the session object back. But we seem to not do that for the origin sessions, and just rely on OpenSSL's own ref counter to clean up things.